### PR TITLE
[expr] Handcode support for equality in the interpreter

### PR DESCRIPTION
**Pull request by bkirwi**
The most interesting function that the `ColumnSpecs` interpreter can't handle yet is `=`... very common but not monotonic.

This PR:
- Breaks out an abstraction for special-cased functions, since we're about to add a second one. (Intent is to make it easier to factor out support for individual functions / keep the special-casing in one place.)
- Add special handling for `=`, along with a couple tests. (It's already allowlisted in the proptest, so we've got some coverage there too.)

### Motivation

Shamed by @philip-stoev's meme:

image

(More seriously: so far we've added support for functions as use-cases come up, and while I've been surprised how rare `=` is for temporal filters, there are definitely good use-cases for it.)

### Tips for reviewer

~I'll do a bit of manual testing as well, but wanted to open up the PR in case anyone has opinions on the restructuring.~ Manual testing is done, and I've expanded some tests for the existing JSON special case to confirm the refactoring didn't break it.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by philip-stoev** [comment]
Thank you, please give me some time to wrap my head around the pecularities of `=` and I will try to apply some randomized end-to-end testing.

**Review by philip-stoev — APPROVED** [review]
No wrong results from randomized testing, so we are good. Thank you.

**Comment by bkirwi** [comment]
Thank your for suggesting and for taking a look!

**Comment by ggevay** [comment]
I'll review this PR today.

File: src/expr/src/interpret.rs [hunk 1]
```diff
@@ -585,6 +585,111 @@ impl From<Monotonic> for Pushdownable {
...
+        fn eq<'b>(left: ResultSpec<'b>, right: ResultSpec<'b>) -> ResultSpec<'b> {
+            eagerly(left, right, |left, right| {
+                // `eq` might return true if there's any overlap between the range of its two arguments...
+                let maybe_true = match left.clone().intersect(right.clone()) {
+                    Values::Empty => ResultSpec::nothing(),
+                    _ => ResultSpec::value(Datum::True),
+                };
+
+                // ...and may return false if the union contains at least two distinct values.
+                let maybe_false = match left.union(right) {
```
Comment (ggevay) [inline src/expr/src/interpret.rs:666]: Maybe add a comment here that if either side is `Values::Empty`, that is handled by `eagerly`.

**Review by ggevay — APPROVED** [review]
LGTM

File: src/expr/src/interpret.rs [hunk 2]
```diff
@@ -633,14 +739,13 @@ impl Interpreter for Trace {
     ) -> Self::Summary {
         // TODO: this is duplicative! If we have more than one or two special-cased functions
```
Comment (aalexandrov) [inline src/expr/src/interpret.rs:740]: nit: I think this TODO can be resolved now.

**Pull request merged**